Dockerfile

```diff
@@ -0,0 +1,39 @@
+# Base on offical Node.js Alpine image
+FROM node:current-alpine AS base
+
+# Set working directory
+WORKDIR /base
+
+# Copy package.json and package-lock.json before other files
+# Utilise Docker cache to save re-installing dependencies if unchanged
+COPY package*.json yarn.lock ./ 
+
+# Install dependencies
+RUN yarn install --frozen-lockfile
+
+# Copy all files
+COPY . .
+
+# Build app
+FROM base AS builder
+ENV NODE_ENV=production
+WORKDIR /build
+COPY . .
+COPY --from=base /base/node_modules ./node_modules
+RUN yarn build
+
+# Production image, copy all the files and run next
+FROM node:current-alpine AS runner
+ENV NODE_ENV=production
+WORKDIR /app
+
+# Copy from build image
+COPY --from=builder /build/package*.json ./
+COPY --from=builder /build/node_modules ./node_modules
+COPY --from=builder /build/.next ./.next
+COPY --from=builder /build/public ./public
```

Comment:
`COPY` them individually, rather than all at once. This ensures that each step’s build successfully. It’s easy to maintain and monitor. I prefer to keep it like this. You can see more examples like [this](https://github.com/vercel/next.js/pull/17794) (please see the file changed)

Reply:
According to your reference, author only have 2 copies while we have 4. This is only a `COPY` command which copies the built file to the `/app` which I still don't think it should be done separately. Not like `RUN` command which some should be separated for monitoring though. What do you think? Please share some thoughts @kimvcp .

Reply:
@LilBank is a balance between running the entire stage even though you only need a section of it, vs having multi-stage where it may be the case you copy things multiple times, but it is optimized for caching the layers.

So technically, depends on which layer of invalidation you are producing.

I would say, test it if you think it is slower, this is the best I came up with https://github.com/vercel/next.js/discussions/16995#discussioncomment-132339

Reply:
Really appreciate your comment 🙇🏻‍♂️. However, I think my [comment](https://github.com/Cobda/cobda-web/pull/17/files#discussion_r575370998) is unclear so let me explain it explicitly.

My first thought was neither of each approach are slower than each other (multi `COPY`s/single `COPY`) and when I tested them the results were as expected. I have this thought because I was trying to follow **Dockerfile best practices** documentation([here](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/)). The document mention that we should **minimize** the number of layers in the image([see here](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers)) since `ADD` and `COPY` command creates a layer. Furthermore, both approach caches will be invalid([here, 3rd bullet point](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache)) since both depends on the **built files** from the `builder` image (file changed => new commit => new build = (cache `COPY =! new COPY`) ).

So this is my thought 😅

In conclusion, both approach gives the same performance but only the generated layers are the diff. What do you think about this? Should we separate it into thin layers of one thick layer? 🤔